exporters/elasticsearch-exporter/src/main/java/io/camunda/zeebe/exporter/ElasticsearchClient.java

```diff
@@ -203,27 +203,70 @@ public boolean putIndexTemplate(
     template.put("composed_of", Collections.singletonList(configuration.index.prefix));
 
-    // update alias in template in case it was changed in configuration
-    final Object templateProperties =
-        template.computeIfAbsent("template", key -> new HashMap<String, Object>());
-    if (templateProperties instanceof Map) {
-      final Map templatePropertyMap = (Map) templateProperties;
-      templatePropertyMap.put(
-          "aliases", Collections.singletonMap(aliasName, Collections.emptyMap()));
+    // update aliases in template in case it was changed in configuration
+    final Map<String, Object> templateProperties = getTemplatePropertyMap(template, templateName);
+    templateProperties.put("aliases", Collections.singletonMap(aliasName, Collections.emptyMap()));
+
+    // update number of shards in template in case it was changed in configuration
+    final Map<String, Object> settingsProperties =
+        getSettingsPropertyMap(templateProperties, templateName);
+    updateSettings(settingsProperties);
+
+    return putIndexTemplate(templateName, template);
+  }
+
+  private Map<String, Object> getTemplatePropertyMap(
+      Map<String, Object> properties, String templateName) {
+    return getProperties(properties, "template", templateName);
+  }
+
+  private Map<String, Object> getSettingsPropertyMap(
+      Map<String, Object> properties, String templateName) {
+    return getProperties(properties, "settings", templateName);
+  }
+
+  private Map<String, Object> getProperties(
+      Map<String, Object> from, String fieldName, String templateName) {
+    final Object properties = from.computeIfAbsent(fieldName, key -> new HashMap<String, Object>());
+    return convertToMap(properties, fieldName, templateName);
+  }
+
+  private void updateSettings(Map<String, Object> settingsProperties) {
+    // update number of shards in template in case it was changed in configuration
+    final Integer numberOfShards = configuration.index.getNumberOfShards();
+    if (numberOfShards != null) {
+      settingsProperties.put("number_of_shards", numberOfShards);
+    }
+
+    // update number of replicas in template in case it was changed in configuration
+    final Integer numberOfReplicas = configuration.index.getNumberOfReplicas();
+    if (numberOfReplicas != null) {
+      settingsProperties.put("number_of_replicas", numberOfReplicas);
+    }
+  }
+
+  private Map<String, Object> convertToMap(
+      final Object obj, final String fieldName, final String templateName) {
+    if (obj instanceof Map) {
+      return (Map<String, Object>) obj;
     } else {
       throw new IllegalStateException(
           String.format(
-              "Expected the 'template' field of the index template '%s' to be a map, but was '%s'",
-              templateName, templateProperties.getClass()));
+              "Expected the '%s' field of the template '%s' to be a map, but was '%s'",
+              fieldName, templateName, obj.getClass()));
```

Comment:
:x: This is not very helpful, because `fieldName` is potentially only the last element in a path.

Reply:
When implementing that part, I had the same thought. However, I decided not to consider this further. Why?

* Implementation-wise, IMO it is (implicitly - but still easy to spot) clear in which cases this method is being called, i.e., when the properties `template` and `settings` are adjusted.
* Looking at the templates, the properties `template` and `settings` are not overloaded with the JSON template. So from a user perspective, when this exception is thrown for example, in the case of the `settings` property, for the user, it will be pretty clear which `settings` property must be fixed as there is only one of `settings` property (the same applies for the `template` property).
* From the current usage, I don't see much benefit in covering that case.

That's why I decided not to put any additional effort into it.

WDYT? Do you see any alternatives?

Reply:
I am a little weary to say "it's fine, because the way it is used now it is clear".

My suggestion would be to rename `templateName` to `context` and rephrase the sentence. Then in the call site you can e.g. compose:
```
...
   final Map<String, Object> settingsProperties =
        getSettingsPropertyMap(templateProperties, templateName + ".template");
...

  private Map<String, Object> getSettingsPropertyMap(
      Map<String, Object> properties, String context) {
    return getProperties(properties, "settings", context + ".settings");
  }
```

Reply:
I applied the changes (almost) as suggested.